include/jemalloc/internal/thread_event.h

```diff
@@ -0,0 +1,113 @@
+#ifndef JEMALLOC_INTERNAL_THREAD_EVENT_H
+#define JEMALLOC_INTERNAL_THREAD_EVENT_H
+
+#include "jemalloc/internal/tsd.h"
+
+/*
+ * Maximum threshold on thread_allocated_threshold_fast, so that there is no
+ * need to check overflow in malloc fast path. (The allocation size in malloc
+ * fast path never exceeds SC_LOOKUP_MAXCLASS.)
+ */
+#define THREAD_ALLOCATED_THRESHOLD_FAST_MAX				\
+    (UINT64_MAX - SC_LOOKUP_MAXCLASS + 1)
+
+/*
+ * List of all waits, in the following format:
+ *  W(event,		(condition))
+ */
+#define ITERATE_OVER_ALL_WAITS						\
+    W(prof_sample,	(config_prof && opt_prof))
+
+void thread_event_trigger(tsd_t *tsd, uint64_t usize, bool next_time);
+void thread_event_update(tsd_t *tsd);
+
+JEMALLOC_ALWAYS_INLINE void
+thread_event_assert_invariants(tsd_t *tsd) {
+	if (!config_debug) {
+		return;
+	}
+
+	uint64_t thread_allocated = tsd_thread_allocated_get(tsd);
+	uint64_t last_event = tsd_thread_allocated_last_event_get(tsd);
+	uint64_t threshold = tsd_thread_allocated_threshold_get(tsd);
+	uint64_t threshold_fast = tsd_thread_allocated_threshold_fast_get(tsd);
+
+	assert(last_event != threshold);
+	if (last_event < threshold &&
+	    threshold <= THREAD_ALLOCATED_THRESHOLD_FAST_MAX) {
+		assert(threshold_fast == threshold);
+	} else {
+		assert(threshold_fast == 0);
+	}
+
+	/* Both subtractions are intentionally susceptible to underflow. */
+	assert(thread_allocated - last_event < threshold - last_event);
+
+	/*
+	 * The rest of this function checks the wait times, and it can only
+	 * fail (trivially) when both of the following two conditions hold:
+	 * (1) the TSD is just initialized, and
+	 * (2) none of the events has its condition on.
+	 * The thread_allocated == 0 guard below is stronger than needed, but
+	 * having an exactly accurate guard is more complicated to implement.
+	 */
+	if (thread_allocated == 0) {
+		return;
+	}
+
+	uint64_t wait = UINT64_MAX;
+
+#define W(event, condition)						\
+	if (condition) {						\
+		uint64_t event_wait =					\
+		    tsd_##event##_event_wait_get(tsd);			\
+		if (event_wait > 0 && event_wait < wait) {		\
+			wait = event_wait;				\
+		}							\
+	}
+
+	ITERATE_OVER_ALL_WAITS
+#undef W
+
+	assert(wait == threshold - last_event);
+}
+
+JEMALLOC_ALWAYS_INLINE bool
+thread_event(tsd_t *tsd, size_t usize) {
+	thread_event_assert_invariants(tsd);
+
+	uint64_t thread_allocated_before = tsd_thread_allocated_get(tsd);
+
+	/*
+	 * Avoid using the setter so that thread_allocated is incremented even
+	 * when tsd state is reincarnated or minimal_initialized.
+	 */
+	*tsd_thread_allocatedp_get(tsd) = thread_allocated_before + usize;
```

Comment:
I thought about that, but this is on medium fast path: if we use `+=`, I worried that the compiler might decide to read from the memory again, since the compiler after all should not assume that the memory was unchanged between the two reads. Meanwhile we know that this is very unlikely and we probably don't care even if it happens(?).

Reply:
On the contrary, if you don’t tell compiler a location may change (e.g. with volatile), it will assume the content won’t be changed. However I think make it explicit is useful too, let’s keep it as is then.